utils/wav2adpcm/wav2adpcm.c

```diff
@@ -266,63 +354,113 @@ int wav2adpcm(const char *infile, const char *outfile) {
         return -1;
     }
 
-    if(validate_wav_header(&wavhdr, 1, 16, in)) {
+    // printf("Header size, %d bytes\n", wavhdr.hdrsize);
+    // printf("Format: %#x\n", wavhdr.format);
+    // printf("Channels: %d\n", wavhdr.channels);
+    // printf("Freq: %d\n", wavhdr.freq);
+    // printf("Bit depth: %d\n", wavhdr.bits_per_sample);
+
+    /* If the input is the desired output format, just copy */
+    if(wavhdr.format == WAVE_FMT_YAMAHA_ADPCM ||
+       wavhdr.format == WAVE_FMT_YAMAHA_ADPCM_ITU_G723) {
+        return straight_copy(in, outfile);
+    }
+
+    if(validate_wav_header(&wavhdr, &wavhdr_chunk, WAVE_FMT_PCM, 16, in)) {
         fclose(in);
         return -1;
     }
 
-    pcmsize = wavhdr.datasize;
-
+    pcmsize = wavhdr_chunk.datasize;
     adpcmsize = pcmsize / 4;
+
     pcmbuf = malloc(pcmsize);
     adpcmbuf = malloc(adpcmsize);
+    if(!pcmbuf || !adpcmbuf) {
+        fprintf(stderr, "Memory allocation failed.\n");
+        result = -1;
+        goto cleanup;
+    }
 
     if(fread(pcmbuf, pcmsize, 1, in) != 1) {
         fprintf(stderr, "Cannot read data.\n");
-        fclose(in);
-        return -1;
+        result = -1;
+        goto cleanup;
     }
     fclose(in);
+    in = NULL;
 
-    if(wavhdr.channels == 1) {
+    if(wavhdr.channels == 1)
         pcm2adpcm(adpcmbuf, pcmbuf, pcmsize);
-    }
     else {
         /* For stereo we just deinterleave the input and store the
            left and right channel of the ADPCM data separately. */
         deinterleave(pcmbuf, pcmsize);
         pcm2adpcm(adpcmbuf, pcmbuf, pcmsize / 2);
-        pcm2adpcm(adpcmbuf + adpcmsize / 2, pcmbuf + pcmsize / 4, pcmsize / 2);
-    }
+        pcm2adpcm(adpcmbuf + adpcmsize / 2, pcmbuf + pcmsize / 4,  pcmsize / 2);
 
-    wavhdr.datasize = adpcmsize;
-    wavhdr.format = 20; /* ITU G.723 ADPCM (Yamaha) */
-    wavhdr.bits = 4;
-    wavhdr.totalsize = wavhdr.datasize + sizeof(wavhdr) - 8;
+        if(interleaved)
+            interleave_adpcm(adpcmbuf, adpcmsize);
+    }
 
     out = fopen(outfile, "wb");
-    if(fwrite(&wavhdr, sizeof(wavhdr), 1, out) != 1 || 
-        fwrite(adpcmbuf, adpcmsize, 1, out) != 1) {
-        fprintf(stderr, "Cannot write ADPCM data.\n");
-        fclose(out);
-        return -1;
+    if(!out) {
+        fprintf(stderr, "Cannot open output file for writing.\n");
+        result = -1;
+        goto cleanup;
     }
-    
-    fclose(out);
 
-    return 0;
+    if(no_header) {
+        /* Just write the body */
+        if(fwrite(adpcmbuf, adpcmsize, 1, out) != 1) {
+            fprintf(stderr, "Cannot write ADPCM data.\n");
+            result = -1;
+            goto cleanup;
+        }
+    }
+    else {
+        /* Build header */
+        wavhdr.hdrsize = 0x10;
+        wavhdr.format = (wavhdr.channels == 2 && interleaved) ? WAVE_FMT_YAMAHA_ADPCM : WAVE_FMT_YAMAHA_ADPCM_ITU_G723;
```

Comment:
```suggestion
        wavhdr.format = interleaved ? WAVE_FMT_YAMAHA_ADPCM : WAVE_FMT_YAMAHA_ADPCM_ITU_G723;
```

Reply:
Need ability for mono WAVE_FMT_YAMAHA_ADPCM.

Reply:
Done